compiler/src/main/kotlin/io/spine/protodata/backend/DependencyEvents.kt

```diff
@@ -0,0 +1,194 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata.backend
+
+import com.google.protobuf.Descriptors.Descriptor
+import com.google.protobuf.Descriptors.EnumDescriptor
+import com.google.protobuf.Descriptors.FileDescriptor
+import com.google.protobuf.Descriptors.ServiceDescriptor
+import io.spine.protodata.EnumType
+import io.spine.protodata.FilePath
+import io.spine.protodata.MessageType
+import io.spine.protodata.ProtobufSourceFile
+import io.spine.protodata.Service
+import io.spine.protodata.enumType
+import io.spine.protodata.event.dependencyDiscovered
+import io.spine.protodata.messageType
+import io.spine.protodata.name
+import io.spine.protodata.oneofGroup
+import io.spine.protodata.path
+import io.spine.protodata.service
+import io.spine.protodata.typeUrl
+
+/**
+ * Creates a `DependencyDiscovered` event from the given file descriptor.
+ *
+ * The event reflects all the definitions from the file.
+ */
+internal fun discoverDependencies(fileDescriptor: FileDescriptor) =
+    dependencyDiscovered {
+        val id = fileDescriptor.path()
+        file = id
+        content = fileDescriptor.toPbSourceFile()
+    }
+
+private fun FileDescriptor.toPbSourceFile(): ProtobufSourceFile {
+    val path = path()
+    val result = ProtobufSourceFile.newBuilder()
+        .setFilePath(path)
+        .setFile(toFileWithOptions())
+    val doc = Documentation.fromFile(this)
+    with(DefinitionsBuilder(path, doc)) {
+        result.putAllType(
+            messageTypes()
+                .map { it.name.typeUrl() to it }
+                .toMap()
+        )
+        result.putAllEnumType(
+            enumTypes()
+                .map { it.name.typeUrl() to it }
+                .toMap()
+        )
+        result.putAllService(
+            services()
+                .map { it.name.typeUrl() to it }
```

Comment:
Please extract the `.map {...}.toMap()`. We have them 3 times already here. Please also look for similar cases and refactor.